mithril-client/src/feedback.rs

```diff
@@ -0,0 +1,568 @@
+//! Then pass an instance of your receiver when building your `Client` using
+//! [`ClientBuilder::add_feedback_receiver`][crate::ClientBuilder::add_feedback_receiver] method.
+//!
+//! # Example
+//!
+//! Using the provided [SlogFeedbackReceiver] to log the events using a [slog] logger.
+//!
+//! ```no_run
+//! use std::sync::Arc;
+//! # async fn run() -> mithril_client::MithrilResult<()> {
+//! use mithril_client::{ClientBuilder, MessageBuilder, feedback::SlogFeedbackReceiver};
+//!
+//! let client = ClientBuilder::aggregator("YOUR_AGGREGATOR_ENDPOINT", "YOUR_GENESIS_VERIFICATION_KEY")
+//!     .add_feedback_receiver(Arc::new(SlogFeedbackReceiver::new(build_logger())))
+//!     .build()?;
+//!
+//! let _ = client.certificate().verify_chain("CERTIFICATE_HASH").await?;
+//! #
+//! #    Ok(())
+//! # }
+//!
+//! pub fn build_logger() -> slog::Logger {
+//!   use slog::Drain;
+//!   let decorator = slog_term::TermDecorator::new().build();
+//!   let drain = slog_term::FullFormat::new(decorator).build().fuse();
+//!   let drain = slog_async::Async::new(drain).build().fuse();
+//!
+//!   slog::Logger::root(Arc::new(drain), slog::o!())
+//! }
+//! ```
+//!
+//! Running this code should yield the following logs (example run on _pre-release-preview_):
+//!
+//! ```shell
+//! Nov 08 14:41:40.436 INFO Certificate chain validation started, certificate_chain_validation_id: ab623989-b0ac-4031-8522-1370958bbb4e
+//! Nov 08 14:41:40.626 INFO Certificate validated, certificate_chain_validation_id: ab623989-b0ac-4031-8522-1370958bbb4e, certificate_hash: dd4d4299cfb817b5ee5987c3de7cf5f13bdcda69c968ef087effd550470dc081
+//! Nov 08 14:42:05.477 INFO Certificate validated, certificate_chain_validation_id: ab623989-b0ac-4031-8522-1370958bbb4e, certificate_hash: 660b3d426a95303254bb255a56bed443616ea63c4d721ea77433b920d7ebdf62
+//! Nov 08 14:42:05.477 INFO Certificate chain validated, certificate_chain_validation_id: ab623989-b0ac-4031-8522-1370958bbb4e
+//! ```
+
+use async_trait::async_trait;
+use slog::{info, Logger};
+use std::sync::{Arc, RwLock};
+use uuid::Uuid;
+
+/// Event that can be reported by a [FeedbackReceiver].
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum MithrilEvent {
+    /// A snapshot download has started
+    SnapshotDownloadStarted {
+        /// Digest of the downloaded snapshot
+        digest: String,
+        /// Unique identifier used to track this specific snapshot download
+        download_id: String,
+        /// Size of the downloaded archive
+        size: u64,
+    },
+    /// A snapshot download is in progress
+    SnapshotDownloadProgress {
+        /// Unique identifier used to track this specific snapshot download
+        download_id: String,
+        /// Number of bytes that have been downloaded
+        downloaded_bytes: u64,
+        /// Size of the downloaded archive
+        size: u64,
+    },
+    /// A snapshot download has completed
+    SnapshotDownloadComplete {
```

Comment:
It would be more natural to call this event like this IMO:
```suggestion
    SnapshotDownloadCompleted {
```